### ISSUE TITLE
speed up CLI commands: significantly reduce calls to this.authenticate()

### ISSUE THREAD

**Issue by sw-yx** [netlify/cli#286]
**- Do you want to request a *feature* or report a *bug*?**

bug

**- What is the current behavior?**

currently we call `this.authenticate()` in many places:
![image](https://user-images.githubusercontent.com/6764957/55672728-64a8b100-586c-11e9-8d81-1fdd5553e2a9.png)

this is an expensive call to `getCurrentUser`, as discussed here https://github.com/netlify/cli/issues/285

However we don't even bother to do this for `netlify login`: https://github.com/netlify/cli/blob/master/src/commands/login.js#L9

this feels wrong. we are taking a perf hit every time we use a netlify command when it is very unlikely to be needed, especially when our login method doesnt even hold itself to a higher standard.

i suggest replacing almost all but the most mission critical `authenticate` calls to just directly retrieving `accessToken` or `configToken`

**Comment by DavidWells** [comment]
The https://github.com/netlify/cli/blob/master/src/base/index.js#L86 call is the validate the token is valid before calling any other API methods

If it’s not, the user is prompted to login if the token is no longer valid or has been revoked.

We could wrap all other API calls in try/catch and prompt for login. This will require a change pretty much everywhere this.netlify.api is used

This would remove the first check and likely speed up the calls. 

One concern would be: developers forgetting to try/catch their calls to Netlify through the sdk and a weird error showing if they aren’t logged in.

But I think it could work... this would result in `this.netlify.api.(whatever)` trying to call the API with found token, catching any failures of token invalid, and then showing a nice error message “auto failed, please Netlify login”

**Comment by sw-yx** [comment]
cc @bcomnes for thoughts.. would this be a breaking change?

**Comment by bcomnes** [comment]
Fixed the link in @DavidWells's post.  

Honestly, the real fix here is having a snappier API.   We could be more optimistic about the token we have anot not check when we look it up from config.

**Comment by bcomnes** [comment]
Can someone unassigned me on this?  I can't for some reason.

**Comment by sw-yx** [comment]
Some perf benchmarks as of v2.11.19:

- `time netlify -v`: `netlify -v  1.48s user 0.21s system 110% cpu 1.521 total`
- `time netlify`: `netlify  1.51s user 0.20s system 112% cpu 1.515 total`
- `time netlify status`: `netlify status  1.57s user 0.22s system 85% cpu 2.099 total`

**Comment by DavidWells** [comment]
Looks like there is something else that is super slow.

`time netlify logout` is also very slow and makes no API calls

**Comment by sw-yx** [comment]
i profiled it once and it seems the autosuggest plugin (aka the "did you mean") takes quite a while to boot up and also blocks for some reason. didnt have time to dig in further. 

@raeesbhatti if you want to work on this, just pointing to some hints

**Comment by DavidWells** [comment]
What are you using for profiling?

---

A couple of things to optimize:

The dev plugin has primarily `sync` fs calls. Need to async those. https://github.com/netlify/netlify-dev-plugin/search?q=Sync&unscoped_q=Sync

These 2 FS read/map runs on every command. They need to run inside of the dev command to only run when they are used.

- https://github.com/netlify/netlify-dev-plugin/blob/master/src/detect-functions-builder.js#L3-L8
- https://github.com/netlify/netlify-dev-plugin/blob/master/src/detect-server.js#L6-L22

You can see these loops running on `netlify logout` for example

![image](https://user-images.githubusercontent.com/532272/61393458-d766f680-a875-11e9-9f17-e3205775e6f6.png)

**Comment by DavidWells** [comment]
A rough baseline I have for "empty oclif commands" speed is `~0.33s`

**Comment by DavidWells** [comment]
This is the logout runtime with `netlify-dev-plugin` completely disabled. 1.18 sec

![image](https://user-images.githubusercontent.com/532272/61394906-0cc11380-a879-11e9-8cd6-afd475a2cf8a.png)

It  appears (some of) the slow downs are from the global FS calls running on every command

**Comment by sw-yx** [comment]
yuk. easy to inline those calls for sure. 

for profiling i was just looking at the output when setting `DEBUG=*` and seeing which calls had the biggest numbers.

**Comment by raeesbhatti** [comment]
It seems like `this.authenticate()` is no longer an issue since we've improved how it works. It works by:
1. checking CLI flag
1. checking ENV var for a token
1. checking global config for a token

And only after all of these fail, do we call `expensivelyAuthenticate()` which then calls `getCurrentUser()` after it has acquired a token.

[https://github.com/netlify/cli-utils/blob/master/src/index.js#L140-L165](https://github.com/netlify/cli-utils/blob/d1f89d6008dda14857c275582d49ee2906867009/src/index.js#L140-L165)

I'll close this issue for now, maybe we can create another issue for general performance improvement.
